src/OrleansAWSUtils/Providers/Storage/DynamoDBStorage.cs

```diff
@@ -0,0 +1,334 @@
+            if (useJsonFormat)
+            {
+                // http://james.newtonking.com/json/help/index.html?topic=html/T_Newtonsoft_Json_JsonConvert.htm
+                entity.StringData = JsonConvert.SerializeObject(grainState, jsonSettings);
+                dataSize = STRING_DATA_PROPERTY_NAME.Length + entity.StringData.Length;
+            }
+            else
+            {
+                // Convert to binary format
+                entity.Data = SerializationManager.SerializeToByteArray(grainState);
+                dataSize = BINARY_DATA_PROPERTY_NAME.Length + entity.Data.Length;
+            }
+
+            var pkSize = PK_PROPERTY_NAME.Length + entity.PartitionKey.Length;
+            var rkSize = RK_PROPERTY_NAME.Length + entity.RowKey.Length;
+            var versionSize = VERSION_PROPERTY_NAME.Length + (entity.Version.HasValue ? entity.Version.Value.ToString().Length : 0);
+
+            if ((pkSize + rkSize + versionSize + dataSize) > MAX_DATA_SIZE)
+            {
+                var msg = string.Format("Data too large to write to DynamoDB table. Size={0} MaxSize={1}", dataSize, MAX_DATA_SIZE);
+                throw new ArgumentOutOfRangeException("GrainState.Size", msg);
+            }
+        }
+
+        private string GetKeyString(GrainReference grainReference)
+        {
+            var key = string.Format("{0}_{1}", serviceId, grainReference.ToKeyString());
+            return AWSUtils.ValidateDynamoDBPartitionKey(key);
+        }
+
+        internal class GrainStateRecord : DynamoDBEntity
+        {
+            public GrainStateRecord()
+            {
+                PartitionKey = string.Empty;
+                RowKey = string.Empty;
+            }
+            public byte[] Data { get; set; }
+
+            public string StringData { get; set; }
+        }
+
+        private class GrainStateTableDataManager
+        {
+            public string TableName { get; private set; }
+            private readonly Logger logger;
+            private readonly DynamoDBDataManager<GrainStateRecord> tableManager;
+
+            public GrainStateTableDataManager(string tableName, string accessKey, string secretKey, string service, int readCapacityUnits, int writeCapacityUnits, Logger logger)
+            {
+                this.logger = logger;
+                TableName = tableName;
+                tableManager = new DynamoDBDataManager<GrainStateRecord>(tableName, accessKey, secretKey, service, readCapacityUnits, writeCapacityUnits, logger);
+            }
+
+            public Task InitTableAsync()
+            {
+                return tableManager.InitTableAsync();
+            }
+
+            public Task<GrainStateRecord> Read(string partitionKey, string rowKey)
+            {
+                return tableManager.ReadSingleTableEntryAsync(partitionKey, rowKey);
+            }
+
+            public Task Write(GrainStateRecord record)
+            {
+                return tableManager.UpsertItemAsync(record);
```

Comment:
Look at https://github.com/dotnet/orleans/blob/master/src/OrleansAzureUtils/Providers/Storage/AzureTableStorage.cs#L496.
There is a good reason it does create OR update, but not upsert.
We want to fail if we try to do create but it already exists. Otherwise, we can overwrite by mistake.